File: Python/Problems/ProblemSets/Problem2-2.py


class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class LinkedList:
    def __init__(self):
        self.head = None


    def push(self, val):
        new_node = ListNode(val)

        if self.head is None:
            self.head = new_node
            return 
        
        last = self.head
        while last.next is not None:
            last = last.next 
        last.next = new_node




    def __str__(self):
        ret_str = '['
        temp = self.head
        while temp is not None:
            ret_str += str(temp.val) + ', '
            temp = temp.next

        ret_str = ret_str.rstrip(', ')
        ret_str += ']'
        return ret_str
    


def hasCycle(head):
    slow = head
    fast = head
    
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
        if slow == fast:
            return True
        
    return False




#for 1st case:

l = LinkedList()
for val in [3,2,0,-4]:
    l.push(val)

l.head.next.next.next.next = l.head.next  #-4 points to 2
print("Cycle present in the linked list, where the tail connects to the 1st node (0-indexed): " , hasCycle(l.head))




# for second case: 

l = LinkedList()
for val in [1,2]:
    l.push(val)

l.head.next = l.head
print("Cycle present  in the linked list, where the tail connects to the 0th node: " , hasCycle(l.head))





#for third case: 
l = LinkedList()
for val in [1]:
    l.push(val)

l.head = l.head
print("There is no cycle in the linked list: " , hasCycle(l.head))












# Unprofessional but easy to understand way 


# # linked list banate hain
# n1 = ListNode(3)
# n2 = ListNode(2)
# n3 = ListNode(0)
# n4 = ListNode(-4)

# for i in [3,2,0,-4]:
#     print (i)

# n1.next = n2
# n2.next = n3
# n3.next = n4
# n4.next = n2  # cycle

# # ab function call
# print("Cycle Present in the linked list: " , hasCycle(n1))  # head = n1

File: Cpp/Problems/ProblemSets/Problem2-2.cpp

#include <iostream>
using namespace std;
// int main()

class Node
{
public:
    int val;
    Node *next;

    Node(int v)
    {
        val = v;
        next = nullptr;
    }
};

class LinkedList
{
public:
    Node *head;
    LinkedList()
    {
        head = nullptr;
    }

    void Push(int val)
    {
        Node *new_node = new Node(val);
        if (head == nullptr)
        {
            head = new_node;
            return;
        }

        Node *last = head;
        while (last->next != nullptr)
        {
            last = last->next;
        }

        last->next = new_node;
    }

    string toString()
    {
        string ret_str = "[";
        Node *temp = head;
        while (temp != nullptr)
        {
            ret_str += to_string(temp->val) + ", ";
            temp = temp->next;
        }

        if (ret_str.size() > 1)
        {
            ret_str.pop_back();
            ret_str.pop_back();
        }

        ret_str += "]";
        return ret_str;
    };

    bool hasCycle()
    {

        Node *slow = head;
        Node *fast = head;
        while (fast != nullptr && fast->next != nullptr)
        {
            slow = slow->next;
            fast = fast->next->next;
            if (slow == fast)
            {
                return true;
            }
        }
        return false;
    }
};

// for val in [3,2,0,-4]:
int main()
{
    LinkedList l;
    int arr[] = {3, 2, 0, -4};
    int size = sizeof(arr) / sizeof(arr[0]);
    for (int i = 0; i < size; i++)
    {
        l.Push(arr[i]);
    }
    cout << "LinkedList: " << l.toString() << endl;
    l.head->next->next->next->next = l.head->next;
    cout << boolalpha;
    cout << "Cycle present in the LinkedList: " << l.hasCycle() << endl;

    // if cycle is not present :
    LinkedList l2;
    l2.Push(1);
    cout << "LinkedList: " << l2.toString() << endl;
    l2.head = l2.head;
    cout << boolalpha;
    cout << "Cycle present in the LinkedList: " << l2.hasCycle() << endl;
    return 0;
};
